kpops/pipeline_generator/pipeline.py

```diff
@@ -229,45 +284,71 @@ def __str__(self) -> str:
             )
         )
 
-    @staticmethod
-    def substitute_component_specific_variables(
-        component: PipelineComponent, pair: dict  # TODO: better parameter name for pair
-    ) -> dict:
-        """Overrides component config with component config in pipeline environment definition."""
-        # HACK: why do we need an intermediate JSON object?
-        component_data: dict = json.loads(
-            substitute(
-                json.dumps(pair),
-                {
-                    "component_type": component.type,
-                    "component_name": component.name,
-                },
+    def substitute_in_component(self, component_as_dict: dict) -> dict:
+        """Substitute all $-placeholders in a component in dict representation
+
+        :param component_as_dict: Component represented as dict
+        :type component_as_dict: dict
+        :return: Updated component
+        :rtype: dict
+        """
+        config = self.config
+        # All variables that were previously introduced in the component by the substitution
+        # functions, still hardcoded.
+        substitution_hardcoded = {
+            # "component_name": component.name,
+            # "component_type": component.type,
+            "error_topic_name": config.topic_name_config.default_error_topic_name,
+            "output_topic_name": config.topic_name_config.default_output_topic_name,
+            # "schema_registry_url": self.config.schema_registry_url,
+            # "broker": self.config.broker,
+        }
+        # Generate a substitution dict from the component
+        component_substitution = generate_substitution(
+            component_as_dict,
+            "component",
+            substitution_hardcoded,
+        )
+        # Generate a substitution dict from the pipeline config and combine it
+        # with the component-specifig substitution
+        substitution = generate_substitution(
+            json.loads(config.json()), "", component_substitution
+        )
+        # Substitute all vars in the component, avoid duplicates, prioritize component def
+        substituted_component: dict = json.loads(
+            substitute_nested(
+                json.dumps(component_as_dict),
+                **update_nested_pair(substitution, os.environ),
```

Comment:
I am not a fan of having so many comments written in the code. This is not a good practice since the comment can get outdated, which leads to misleading information. If you think your method is too long and hard to understand each part, break each part into, smaller methods with good names, and docs.

Reply:
I agree with you, I planned on removing them after the reviews. I expected that I wouldn't do too well explaining my solution otherwise.

Reply:
Done